openwhisk/executor.go

```diff
@@ -68,17 +95,160 @@ func NewExecutor(logout *os.File, logerr *os.File, command string, env map[strin
+		}
+
+		proc.remoteLogger = &httpLogger{
+			url:     "https://logs.collector.solarwinds.com/v1/log",
+			headers: map[string]string{"Authorization": fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(":"+env[papertrailTokenEnv])))},
+			http:    http.DefaultClient,
+			format:  FormatLogtail, // TODO: Is there a better format for papertrail?
+		}
+	case "datadog":
+		if env[datadogSiteEnv] == "" || env[datadogApiKeyEnv] == "" {
+			return fmt.Errorf("%q and %q have to be an environment variable of the action", datadogSiteEnv, datadogApiKeyEnv)
+		}
+
+		proc.remoteLogger = &httpLogger{
+			url:     fmt.Sprintf("https://http-intake.logs.%s/api/v2/logs", env[datadogSiteEnv]),
+			headers: map[string]string{"DD-API-KEY": env[datadogApiKeyEnv]},
+			http:    http.DefaultClient,
+			format:  FormatDatadog,
+		}
 	}
+	return nil
+}
+
+// consumeStream consumes a log stream into the given channel.
+func consumeStream(streamName string, stream io.Reader, ch chan LogLine) error {
+	s := bufio.NewScanner(stream)
+	for s.Scan() {
+		ch <- LogLine{
+			Time:    time.Now(),
+			Message: s.Text(),
+			Stream:  streamName,
+		}
+	}
+	return s.Err()
 }
 
 // Interact interacts with the underlying process
 func (proc *Executor) Interact(in []byte) ([]byte, error) {
-	// input to the subprocess
+	if proc.lines == nil {
+		// Remote logging is not configured. Just roundtrip and return immediately.
+		return proc.roundtrip(in)
+	}
+
+	// Fetch metadata from the incoming parameters
+	var metadata activationMetadata
+	if err := json.Unmarshal(in, &metadata); err != nil {
+		return nil, fmt.Errorf("failed to parse activation metadata: %w", err)
+	}
+
+	var grp errgroup.Group
+	grp.Go(func() error {
+		var sawStdoutSentinel, sawStdErrSentinel bool
+		var lastErr error
+		for line := range proc.lines {
+			if line.Message == logSentinel {
+				if line.Stream == "stdout" {
+					sawStdoutSentinel = true
+				} else {
+					sawStdErrSentinel = true
+				}
+				if sawStdoutSentinel && sawStdErrSentinel {
+					// We've seen both sentinels. Stop consuming until the next request.
+					return lastErr
+				}
+				continue
+			}
+
+			line.ActivationId = metadata.ActivationId
+			line.ActionName = metadata.ActionName
+			if err := proc.remoteLogger.Send(line); err != nil {
+				// We want to continue consuming all of the logs so we don't exit immediately
+				// but surface the error eventually.
+				lastErr = fmt.Errorf("failed to send log to remote location: %w", err)
+			}
+		}
+		return lastErr
+	})
+
+	proc.logout.WriteString(remoteLogNotice)
+	proc.logout.WriteString("\n")
+	out, err := proc.roundtrip(in)
+
+	// Wait for the streams to process completely.
+	if err := grp.Wait(); err != nil {
+		fmt.Fprintf(proc.logerr, "Failed to process logs: %v\n", err)
+	}
+
+	if err := proc.remoteLogger.Flush(); err != nil {
+		fmt.Fprintf(proc.logerr, "Failed to flush logs to remote location: %v\n", err)
```

Comment:
same comment as above.

Reply:
I saw that `Flush` is not implemented, should we remove this from here, seems like unused code?

Reply:
Ack, will reintroduce it once it makes sense.